include/bluetooth/services/ots.h

```diff
@@ -0,0 +1,177 @@
+			/** Bit 4 Appending data to this object that increases
+			 *	  its Allocated Size is permitted
+			 */
+			bool append_permitted	: 1;
+			/** Bit 5 Truncation of this object is permitted */
+			bool truncate_permitted	: 1;
+			/** Bit 6 Patching this object by overwriting some
+			 *	  of the object's existing contents is permitted
+			 */
+			bool patch_permitted	: 1;
+			/** Bit 7 This object is a marked object */
+			bool is_marked		: 1;
+		} decoded;
+
+		/* Properties as raw value */
+		u32_t raw;
+	};
+};
+
+/**@brief Descriptor for OTS object intialization. */
+struct bt_gatt_ots_obj_init {
+	/* Object Name */
+	char                        *name;
+
+	/* Object Type */
+	struct bt_gatt_ots_obj_type  type;
+
+	/* Object Size (allocated and current) */
+	u32_t                        size;
+
+	/* Object Properties */
+	struct bt_gatt_ots_obj_props props;
+};
+
+/** @brief OTS server callback structure. */
+struct bt_gatt_ots_cb {
+	/** Object created callback
+	 *
+	 *  This callback is mandatory. It is called whenever a new object is
+	 *  created. object data should be set within the context of this
+	 *  callback.
+	 *
+	 *  @param conn The connection that is requesting object creation or
+	 *              NULL if object is created by bt_gatt_ots_object_add()
+	 *  @param id   Object ID
+	 *  @param data Pointer to receive the buffer to object data
+	 *  @param size Allocated Size of the object
+	 *
+	 *  @return 0 in case of success or negative value in case of error.
+	 *  Possible return values:
+	 *  -ENOMEM if no available space for new object.
+	 */
+	int (*obj_created)(struct bt_conn *conn, u64_t id, u8_t **data,
+			   u32_t size);
+
+	/** Object deleted callback
+	 *
+	 *  This callback is called whenever an object is deleted.
+	 *
+	 *  @param conn The connection that deleted the object or NULL if
+	 *              this request came from the server
+	 *  @param id   Object ID
+	 *  @param data Object data
+	 */
+	void (*obj_deleted)(struct bt_conn *conn, u64_t id, u8_t *data);
+
+	/** Object selected callback
+	 *
+	 *  This callback is called on successful object selection.
+	 *
+	 *  @param conn The connection that selected new object
+	 *  @param id   Object ID
+	 */
+	void (*obj_selected)(struct bt_conn *conn, u64_t id);
+
+	/** Object read callback
+	 *
+	 *  This callback is called on successful object read.
+	 *
+	 *  @param conn The connection that read object
+	 *  @param id   Object ID
+	 */
+	void (*obj_read)(struct bt_conn *conn, u64_t id);
+};
+
+/** @brief Add an object to the OTS Server.
+ *
+ *  This function adds an object to the OTS database. When the
+ *  object is being added, a callback obj_created() is called
+ *  to notify the user about new object ID and to receive pointer
+ *  to object data.
+ *
+ *  @param obj_init Meta data of the object
+ *
+ *  @return 0 in case of success or negative value in case of error.
+ */
+int bt_gatt_ots_object_add(struct bt_gatt_ots_obj_init *obj_init);
```

Comment:
I'm not sure what our goal with the API here is: Is it something that we expect to keep in future updates of OTS, or would it be okay to update the API significantly when we e.g. add support for multiple instances? 

The OTS I'm working on atm will have support for multiple instances out-of-the-box and in the API. Basically any function in the API starts with a `struct ots_svc_inst_t *inst` pointer. Even if we only allow a single instance for this implementation, we could have the instance pointer (or an index) as part of the API. 

In brief: Should we design this API to support multiple instances, or change it later?

Reply:
I can prepare API for multi-instance support in the way that you suggested.

However, it's not possible currently to declare OTS instance by the user. I decided to keep the instance descriptor internal (ots_internal.h) to reduce information noise in the public API header. To keep the API as it is, I suggest adding a new API for assigning OTS instance:

```struct bt_gatt_ots *bt_gatt_ots_instance_assign(void)```

Alternatively, I could move all definitions that are required for OTS instance from internal headers to the public ones.

In my opinion, the first option is more preferable. What do you think? Do you foresee any obstacles that would make us fall back to the 2nd option in the future?